File: Complementar - Estruturas de dados (Didático)/Algoritmos22/Alg22.php
<?php
/*
♦ Algoritmo 22: Estrutura de Dados Não Linear Estática (não Dinâmica) de Árvore Binária: Árvore AVL.
*/

class Node {
    public $data;
    public $left;
    public $right;
    public $height;

    public function __construct($data) {
        $this->data = $data;
        $this->left = null;
        $this->right = null;
        $this->height = 1;
    }
}

class AVLTree {
    private $root;

    public function __construct() {
        $this->root = null;
    }

    private function height($node) {
        if ($node === null)
            return 0;
        return $node->height;
    }

    private function balanceFactor($node) {
        if ($node === null)
            return 0;
        return $this->height($node->left) - $this->height($node->right);
    }

    private function rotateRight($y) {
        $x = $y->left;
        $T2 = $x->right;

        $x->right = $y;
        $y->left = $T2;

        $y->height = max($this->height($y->left), $this->height($y->right)) + 1;
        $x->height = max($this->height($x->left), $this->height($x->right)) + 1;

        return $x;
    }

    private function rotateLeft($x) {
        $y = $x->right;
        $T2 = $y->left;

        $y->left = $x;
        $x->right = $T2;

        $x->height = max($this->height($x->left), $this->height($x->right)) + 1;
        $y->height = max($this->height($y->left), $this->height($y->right)) + 1;

        return $y;
    }

    public function insert($data) {
        $this->root = $this->insertNode($this->root, $data);
    }

    private function insertNode($node, $data) {
        if ($node === null)
            return new Node($data);

        if ($data < $node->data)
            $node->left = $this->insertNode($node->left, $data);
        elseif ($data > $node->data)
            $node->right = $this->insertNode($node->right, $data);
        else
            return $node; // Duplicatas não são permitidas

        $node->height = 1 + max($this->height($node->left), $this->height($node->right));

        $balance = $this->balanceFactor($node);

        // Casos de desequilíbrio
        if ($balance > 1) {
            if ($data < $node->left->data) {
                // Rotação simples à direita
                return $this->rotateRight($node);
            } else {
                // Rotação dupla à esquerda-direita
                $node->left = $this->rotateLeft($node->left);
                return $this->rotateRight($node);
            }
        }

        if ($balance < -1) {
            if ($data > $node->right->data) {
                // Rotação simples à esquerda
                return $this->rotateLeft($node);
            } else {
                // Rotação dupla à direita-esquerda
                $node->right = $this->rotateRight($node->right);
                return $this->rotateLeft($node);
            }
        }

        return $node;
    }

    public function remove($data) {
        $this->root = $this->removeNode($this->root, $data);
    }

    private function removeNode($node, $data) {
        if ($node === null)
            return $node;

        if ($data < $node->data)
            $node->left = $this->removeNode($node->left, $data);
        elseif ($data > $node->data)
            $node->right = $this->removeNode($node->right, $data);
        else {
            if ($node->left === null || $node->right === null) {
                $temp = ($node->left !== null) ? $node->left : $node->right;

                if ($temp === null) {
                    $temp = $node;
                    $node = null;
                } else {
                    $node = $temp;
                }
            } else {
                $temp = $this->findMin($node->right);
                $node->data = $temp->data;
                $node->right = $this->removeNode($node->right, $temp->data);
            }
        }

        if ($node === null)
            return $node;

        $node->height = 1 + max($this->height($node->left), $this->height($node->right));

        $balance = $this->balanceFactor($node);

        // Casos de desequilíbrio
        if ($balance > 1) {
            if ($this->balanceFactor($node->left) >= 0) {
                // Rotação simples à direita
                return $this->rotateRight($node);
            } else {
                // Rotação dupla à esquerda-direita
                $node->left = $this->rotateLeft($node->left);
                return $this->rotateRight($node);
            }
        }

        if ($balance < -1) {
            if ($this->balanceFactor($node->right) <= 0) {
                // Rotação simples à esquerda
                return $this->rotateLeft($node);
            } else {
                // Rotação dupla à direita-esquerda
                $node->right = $this->rotateRight($node->right);
                return $this->rotateLeft($node);
            }
        }

        return $node;
    }

    private function findMin($node) {
        while ($node->left !== null)
            $node = $node->left;
        return $node;
    }

    public function inorder() {
        $this->inorderTraversal($this->root);
        echo PHP_EOL;
    }

    private function inorderTraversal($node) {
        if ($node !== null) {
            $this->inorderTraversal($node->left);
            echo $node->data . " ";
            $this->inorderTraversal($node->right);
        }
    }
}

$tree = new AVLTree();

$tree->insert(10);
$tree->insert(20);
$tree->insert(30);
$tree->insert(40);
$tree->insert(50);

echo "Árvore AVL In-Order:" . PHP_EOL;
$tree->inorder();

$tree->remove(30);
$tree->remove(40);

echo "Árvore AVL In-Order após a remoção de 30 e 40:" . PHP_EOL;
$tree->inorder();

?>

File: Complementar - Estruturas de dados (Didático)/Algoritmos22/Alg22.rb
=begin
♦ Algoritmo 22: Estrutura de Dados Não Linear Estática (não Dinâmica) de Árvore Binária: Árvore AVL.
=end

class Node
    attr_accessor :data, :left, :right, :height
  
    def initialize(data)
      @data = data
      @left = nil
      @right = nil
      @height = 1
    end
end
  
class AVLTree
    def initialize
      @root = nil
    end
  
    def height(node)
      return 0 if node.nil?
  
      node.height
    end
  
    def balance_factor(node)
      return 0 if node.nil?
  
      height(node.left) - height(node.right)
    end
  
    def rotate_right(y)
      x = y.left
      t2 = x.right
  
      x.right = y
      y.left = t2
  
      y.height = [height(y.left), height(y.right)].max + 1
      x.height = [height(x.left), height(x.right)].max + 1
  
      x
    end
  
    def rotate_left(x)
      y = x.right
      t2 = y.left
  
      y.left = x
      x.right = t2
  
      x.height = [height(x.left), height(x.right)].max + 1
      y.height = [height(y.left), height(y.right)].max + 1
  
      y
    end
  
    def insert(data)
      @root = insert_node(@root, data)
    end
  
    def insert_node(node, data)
      return Node.new(data) if node.nil?
  
      if data < node.data
        node.left = insert_node(node.left, data)
      elsif data > node.data
        node.right = insert_node(node.right, data)
      else
        return node # Duplicatas não são permitidas
      end
  
      node.height = [height(node.left), height(node.right)].max
  
      balance = balance_factor(node)
  
      # Casos de desequilíbrio
      if balance > 1
        if data < node.left.data
          # Rotação simples à direita
          return rotate_right(node)
        else
          # Rotação dupla à esquerda-direita
          node.left = rotate_left(node.left)
          return rotate_right(node)
        end
      end
  
      if balance < -1
        if data > node.right.data
          # Rotação simples à esquerda
          return rotate_left(node)
        else
          # Rotação dupla à direita-esquerda
          node.right = rotate_right(node.right)
          return rotate_left(node)
        end
      end
  
      node
    end
  
    def remove(data)
      @root = remove_node(@root, data)
    end
  
    def remove_node(node, data)
      return node if node.nil?
  
      if data < node.data
        node.left = remove_node(node.left, data)
      elsif data > node.data
        node.right = remove_node(node.right, data)
      else
        if node.left.nil? || node.right.nil?
          temp = node.left.nil? ? node.right : node.left
  
          return temp.nil? ? node : temp
        else
          temp = find_min(node.right)
          node.data = temp.data
          node.right = remove_node(node.right, temp.data)
        end
      end
  
      return node if node.nil?
  
      node.height = [height(node.left), height(node.right)].max
  
      balance = balance_factor(node)
  
      # Casos de desequilíbrio
      if balance > 1
        if balance_factor(node.left) >= 0
          # Rotação simples à direita
          return rotate_right(node)
        else
          # Rotação dupla à esquerda-direita
          node.left = rotate_left(node.left)
          return rotate_right(node)
        end
      end
  
      if balance < -1
        if balance_factor(node.right) <= 0
          # Rotação simples à esquerda
          return rotate_left(node)
        else
          # Rotação dupla à direita-esquerda
          node.right = rotate_right(node.right)
          return rotate_left(node)
        end
      end
  
      node
    end
  
    def find_min(node)
      while node.left
        node = node.left
      end
      node
    end
  
    def inorder
      inorder_traversal(@root)
      puts
    end
  
    def inorder_traversal(node)
      return if node.nil?
  
      inorder_traversal(node.left)
      print "#{node.data} "
      inorder_traversal(node.right)
    end
end
  
if __FILE__ == $PROGRAM_NAME
    tree = AVLTree.new
  
    tree.insert(10)
    tree.insert(20)
    tree.insert(30)
    tree.insert(40)
    tree.insert(50)
  
    puts "Árvore AVL In-Order:"
    tree.inorder
  
    tree.remove(30)
    tree.remove(40)
  
    puts "Árvore AVL In-Order após a remoção de 30 e 40:"
    tree.inorder
end
  